python/ray/tune/pbt.py

```diff
@@ -0,0 +1,174 @@
+                )
+                
+                #mutate the hyperparamters the new trial, last index of config is always the "env"
+                #so exclude that, temporarily we "perturb" each hyperparameter by a factor 1.2 or 0.8 and then round it to the nearest
+                #even integer
+                print(new_trial.config)
+                new_trial.experiment_tag = trial.experiment_tag + "-mutated@" + str(time)
+                #add the new trial to the trialrunner, and then stop the
+                #  trial it replaced
+                trial_runner._launch_trial(new_trial)
+                trial_runner.add_trial(new_trial)
+                self._results[new_trial] = self._results[trial]
+                self._time_tracker[new_trial] = self._time_tracker[trial]
+                self._results.pop(trial)
+                self._time_tracker.pop(trial)
+                best_trial.stop() if best_trial_was_stopped else print("best trial continues")
+                return TrialScheduler.STOP
+                
+
+
+        else:
+            print("not ready")
+        
+        return TrialScheduler.CONTINUE
+
+    def on_trial_complete(self, trial_runner, trial, result):
+        self._results[trial].append(result)
+        self._completed_trials.add(trial)
+        print("finished a trial")
+        min_time = 0
+        best_trial = None
+        for trial in self._completed_trials:
+            last_result = self._results[trial][-1]
+            if getattr(last_result, self._time_attr) < min_time or min_time == 0:
+                min_time = getattr(last_result, self._time_attr)
+                best_trial = trial
+        print("The Best Trial is currently {0} finishing in {1} iterations, with the hyperparameters of {2}".format(best_trial, min_time, best_trial.config))
+
+    def _default_ready(self, result, trial, time, time_diff = 4):
+        #function checks if trial has been added to the time tracker, if not adds
+        #it, else checks if an appropriate amount of time has passed since the last mutation
+        if trial not in self._time_tracker:
+            print("added trial to time tracker")
+            self._time_tracker[trial] = (time)
+        else:
+            time_since_last = time - self._time_tracker[trial]
+            if time_since_last >= time_diff:
+                self._time_tracker[trial] = time
+                return True
+        return False
+    
+    def _get_best_trial(self, result, time):
+        results_at_time = {}
+        for trial in self._results:
+            results_at_time[trial] = [getattr(r, self._reward_attr)
+                for r in self._results[trial] if getattr(r, self._time_attr) <= time]
+        print("Results at {0}: {1}".format(time, results_at_time))
+        return max(results_at_time, key=lambda x: max(results_at_time.get(x) if results_at_time.get(x) else [0]))
+
+    def _round_up_to_even(self, n):
+        return math.ceil(n/2.) * 2
+
+    def _is_empty(self, x):
```

Comment:
I used it in the updating of hyperparameters, so that were not floats, but considering all hyperparameters aren't numerical, will probably remove this